instrumentation/servlet/servlet-3.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/servlet/v3_0/snippet/SnippetInjectingPrintWriter.java

```diff
@@ -0,0 +1,44 @@
+/* SPDX-License-Identifier: Apache-2.0 */
+package io.opentelemetry.javaagent.instrumentation.servlet.v3_0.snippet;
+
+import java.io.PrintWriter;
+
+public class SnippetInjectingPrintWriter extends PrintWriter {
+  private final String snippet;
+  private final InjectionObject obj = new InjectionObject();
+
+  public SnippetInjectingPrintWriter(PrintWriter writer, String snippet, String characterEncoding) {
+    super(writer);
+    obj.characterEncoding = characterEncoding;
+    obj.headTagBytesSeen = -1;
+    this.snippet = snippet;
+  }
+
+  @Override
+  public void write(String s, int off, int len) {
+    for (int i = off; i < s.length() && i - off < len; i++) {
+      char b = s.charAt(i);
+      write(b);
+    }
```

Comment:
// TODO optimize by handling similar to `ServletOutputStreamInjectionHelper.stringInjection`